koku/api/iam/models.py

```diff
@@ -62,8 +86,101 @@ class Meta:
 class Tenant(TenantMixin):
     """The model used to create a tenant schema."""
 
+    # Sometimes the Tenant model can seemingly return funky results,
+    # so the template schema name is going to get more inline with the
+    # customer account schema names
+    _TEMPLATE_SCHEMA = os.environ.get("TEMPLATE_SCHEMA", "template0")
+    _CLONE_SCHEMA_FUNC_FILENAME = os.path.join(find_db_functions_dir(), "clone_schema.sql")
+    _CLONE_SCHEMA_FUNC_SCHEMA = "public"
+    _CLONE_SHEMA_FUNC_NAME = "clone_schema"
+    _CLONE_SCHEMA_FUNC_SIG = (
+        f"{_CLONE_SCHEMA_FUNC_SCHEMA}.{_CLONE_SHEMA_FUNC_NAME}("
+        "source_schema text, dest_schema text, "
+        "copy_data boolean DEFAULT false, "
+        "_verbose boolean DEFAULT false"
+        ")"
+    )
+
     # Override the mixin domain url to make it nullable, non-unique
     domain_url = None
 
     # Delete all schemas when a tenant is removed
     auto_drop_schema = True
+
+    def _check_clone_func(self):
+        LOG.info(f'Verify that clone function "{self._CLONE_SCHEMA_FUNC_SIG}" exists')
+        res = dbfunc_exists(
+            conn, self._CLONE_SCHEMA_FUNC_SCHEMA, self._CLONE_SHEMA_FUNC_NAME, self._CLONE_SCHEMA_FUNC_SIG
+        )
+        if not res:
+            LOG.warning(f'Clone function "{self._CLONE_SCHEMA_FUNC_SIG}" does not exist')
+            LOG.info(f'Creating clone function "{self._CLONE_SCHEMA_FUNC_SIG}"')
+            apply_sql_file(conn.schema_editor(), self._CLONE_SCHEMA_FUNC_FILENAME, literal_placeholder=True)
+            res = dbfunc_exists(
+                conn, self._CLONE_SCHEMA_FUNC_SCHEMA, self._CLONE_SHEMA_FUNC_NAME, self._CLONE_SCHEMA_FUNC_SIG
+            )
+        return res
+
+    def _verify_template(self, verbosity=1):
+        LOG.info(f'Verify that template schema "{self._TEMPLATE_SCHEMA}" exists')
+        # This is using the teanant table data as the source of truth which can be dangerous.
+        # If this becomes unreliable, then the database itself should be the source of truth
+        # and extra code must be written to handle the sync of the table data to the state of
+        # the database.
+        template_schema = self.__class__.objects.get_or_create(schema_name=self._TEMPLATE_SCHEMA)
+
+        # Strict check here! Both the record and the schema *should* exist!
+        return template_schema and schema_exists(self._TEMPLATE_SCHEMA)
+
+    def _clone_schema(self):
+        result = None
+        with conn.cursor() as cur:
+            # This db func will clone the schema objects
+            # bypassing the time it takes to run migrations
+            sql = """
+select public.clone_schema(%s, %s, copy_data => true) as "clone_result";
+"""
+            LOG.info(f'Cloning template schema "{self._TEMPLATE_SCHEMA}" to "{self.schema_name}" with data')
+            cur.execute(sql, [self._TEMPLATE_SCHEMA, self.schema_name])
+            result = cur.fetchone()
```

Comment:
This could raise a few different exceptions, if the connection or cursor are broken, or if fetchone() returns something other than one result.

Do we need to catch any of those exceptions?

Reply:
If the connection or cursor are broken, then there's precious little that can be done to gracefully handle the condition.
I have placed python checks around the call of this function. If it fails, then the all further DB transaction traffic will fail until it is closed. 
If any exception is thrown from the DB I'd rather it bubble up through the existing infrastructure.

Reply:
Also the function will return only a boolean. Anything else will be an exception from the database that will be picked up via psycopg2 and wrapped by django.